src/main/java/net/earthcomputer/clientcommands/command/WaypointCommand.java

```diff
@@ -0,0 +1,413 @@
+                source.sendFeedback(Component.translatable("commands.cwaypoint.list.empty"));
+                return 0;
+            }
+
+            worldWaypoints.forEach((name, waypoint) -> source.sendFeedback(Component.translatable("commands.cwaypoint.list", name, formatCoordinates(waypoint.location()), waypoint.dimension().location())));
+            return worldWaypoints.size();
+        }
+
+        if (waypoints.isEmpty()) {
+            source.sendFeedback(Component.translatable("commands.cwaypoint.list.empty"));
+            return 0;
+        }
+
+        AtomicInteger count = new AtomicInteger();
+        waypoints.forEach((worldIdentifier, worldWaypoints) -> {
+            if (worldWaypoints.isEmpty()) {
+                return;
+            }
+
+            count.addAndGet(worldWaypoints.size());
+
+            source.sendFeedback(Component.literal(worldIdentifier).append(":"));
+            worldWaypoints.forEach((name, waypoint) -> source.sendFeedback(Component.translatable("commands.cwaypoint.list", name, formatCoordinates(waypoint.location()), waypoint.dimension().location())));
+        });
+        return count.get();
+    }
+
+    private static void saveFile() throws CommandSyntaxException {
+        try {
+            CompoundTag rootTag = new CompoundTag();
+            rootTag.putInt("DataVersion", SharedConstants.getCurrentVersion().getDataVersion().getVersion());
+            CompoundTag compoundTag = new CompoundTag();
+            waypoints.forEach((worldIdentifier, worldWaypoints) -> compoundTag.put(worldIdentifier, worldWaypoints.entrySet().stream()
+                .collect(CompoundTag::new, (result, entry) -> {
+                    CompoundTag waypoint = new CompoundTag();
+                    Tag pos = NbtUtils.writeBlockPos(entry.getValue().location());
+                    waypoint.put("pos", pos);
+                    String dimension = entry.getValue().dimension().location().toString();
+                    waypoint.putString("Dimension", dimension);
+                    result.put(entry.getKey(), waypoint);
+                }, CompoundTag::merge)));
+            rootTag.put("Waypoints", compoundTag);
+            Path newFile = Files.createTempFile(ClientCommands.configDir, "waypoints", ".dat");
+            NbtIo.write(rootTag, newFile);
+            Path backupFile = ClientCommands.configDir.resolve("waypoints.dat_old");
+            Path currentFile = ClientCommands.configDir.resolve("waypoints.dat");
+            Util.safeReplaceFile(currentFile, newFile, backupFile);
+        } catch (IOException e) {
+            throw SAVE_FAILED_EXCEPTION.create();
+        }
+    }
+
+    private static void loadFile() throws Exception {
+        waypoints.clear();
+        CompoundTag rootTag = NbtIo.read(ClientCommands.configDir.resolve("waypoints.dat"));
+        if (rootTag == null) {
+            return;
+        }
+        // TODO: update-sensitive: apply custom data fixes when it becomes necessary
```

Comment:
I'm trying to remove these `// TODO: update-sensitive` comments nowadays and replace them with tests that can be run automatically. But I'll allow you to be lazy here and delete the comment without adding a test, I really think this file format is the least likely out of all the things we have which could silently break.

Reply:
Why not keep the comment? Better than nothing right?